Examples/c++/Linked List/DLList.h

```diff
@@ -0,0 +1,401 @@
+			ASSERT(m_Ptr != m_SentinelNode, "Cannot increment end iterator!");
+			m_Ptr = m_Ptr->Next;
+			return *this;
+		}
+
+		DLLIterator operator++(int)
+		{
+			DLLIterator<List> iterator = *this;
+			++(*this);
+			return iterator;
+		}
+
+		DLLIterator& operator--()
+		{
+			ASSERT(m_Ptr->Previous != m_SentinelNode, "Cannot decrement begin iterator!");
+			m_Ptr = m_Ptr->Previous;
+			return *this;
+		}
+
+		DLLIterator operator--(int)
+		{
+			DLLIterator<List> iterator = *this;
+			--(*this);
+			return iterator;
+		}
+
+		bool operator==(const DLLIterator& rhs) const
+		{
+			return m_Ptr == rhs.m_Ptr;
+		}
+
+		bool operator!=(const DLLIterator& rhs) const
+		{
+			return !(*this == rhs);
+		}
+
+		operator NodePtr()
+		{
+			return m_Ptr;
+		}
+	private:
+		NodePtr m_Ptr;
+		NodePtr m_SentinelNode;
+	};
+
+
+	template<typename T>
+	struct DLNode
+	{
+		using NodePtr = DLNode<T>*;
+
+		T Data = T();
+		NodePtr Next = nullptr;
+		NodePtr Previous = nullptr;
+
+		DLNode() = default;
+		DLNode(const T& data, NodePtr prev, NodePtr next)
+			: Data(data), Previous(prev), Next(next) {}
+		DLNode(T&& data, NodePtr prev, NodePtr next)
+			: Data(std::move(data)), Previous(prev), Next(next) {}
+	};
+
+	template<typename T>
+	class DLList
+	{
+	public:
+		using ValueType = T;
+		using Node = DLNode<T>;
+		using NodePtr = DLNode<T>*;
+		using Iterator = DLLIterator<DLList<T>>;
+	public:
+		DLList()
+		{
+			m_SentinelNode = new Node();
+			m_SentinelNode->Next = m_SentinelNode;
+			m_SentinelNode->Previous = m_SentinelNode;
+		}
+
+		~DLList()
+		{
+			Clear();
+			delete m_SentinelNode;
+		}
+
+		void PushFront(const T& data)
+		{
+			InsertAfter(m_SentinelNode, data);
+		}
+
+		void PushFront(T&& data)
+		{
+			InsertAfter(m_SentinelNode, std::move(data));
+		}
+
+		template<typename... Args>
+		T& EmplaceFront(Args&&... args)
+		{
+			return EmplaceAfter(m_SentinelNode, std::forward<Args>(args)...);
+		}
+
+		void PushBack(const T& data)
+		{
+			InsertBefore(m_SentinelNode, data);
+		}
+
+		void PushBack(T&& data)
+		{
+			InsertBefore(m_SentinelNode, std::move(data));
+		}
+
+		template<typename... Args>
+		T& EmplaceBack(Args&&... args)
+		{
+			return EmplaceBefore(m_SentinelNode, std::forward<Args>(args)...);
+		}
+
+		void Insert(Iterator where, const T& data)
+		{
+			InsertBefore((NodePtr)where, data);
+		}
+
+		void Insert(Iterator where, T&& data)
+		{
+			InsertBefore((NodePtr)where, std::move(data));
+		}
+
+		template<typename... Args>
+		T& Emplace(Iterator where, Args&&... args)
+		{
+			return EmplaceBefore((NodePtr)where, std::forward<Args>(args)...);
+		}
+
+		void PopFront()
+		{
+			DeleteNode(m_SentinelNode->Next);
+		}
+
+		void PopBack()
+		{
+			ASSERT(!Empty(), "PopBack() called on an empty list!");
+			DeleteNode(m_SentinelNode->Previous);
+		}
+
+		void Erase(Iterator where)
+		{
+			ASSERT((NodePtr)where != m_SentinelNode, "Invalid Iterator");
+			DeleteNode(where);
+		}
+
+		void MoveToBack(Iterator from, Iterator to)
```

Comment:
It's not a standart function. Its not very clear why it's included at the list and also the name is not biving the reader perfect understanding what the function is going to do. The minimum for this function is to add a comment to it.

Reply:
I left it there as a utility function but it can be safely removed from the list.